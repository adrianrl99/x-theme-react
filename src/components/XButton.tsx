import styled from '@emotion/styled'

import { XButtonProps } from '~/typing'

const XButton = styled.button<XButtonProps>`
  background-color: ${props =>
    props.color
      ? props.theme.color[props.color]
      : props.theme.palette.white.normal};
  color: ${props => props.theme.palette.white.contrastText};

  cursor: pointer;
  padding: ${props => props.theme.functions.spacing([2, 5])};
  border-radius: ${props => props.theme.functions.spacing([2])};
  font-size: ${props => props.theme.functions.spacing([4])};
  text-transform: uppercase;
  text-align: center;

  ${props => props.theme.mixins.boxShadow(props.theme.functions.shadow(0))}
  ${props =>
    props.theme.mixins.transition(
      props.theme.functions.transition(['transform'], 'fast'),
    )}
  position: relative;

  &:before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    ${props =>
      props.theme.mixins.transition(
        props.theme.functions.transition(['background-color'], 'fast'),
      )}
  }

  &:hover {
    ${props => props.theme.mixins.transform('scale(1.08)')}
    &:before {
      background-color: rgba(0, 0, 0, 0.15);
    }
  }

  &:active {
    ${props =>
      props.theme.mixins.transition(
        props.theme.functions.transition(['transform'], 'faster'),
      )}
    ${props => props.theme.mixins.transform('scale(0.95)')}
  }
`

export default XButton
