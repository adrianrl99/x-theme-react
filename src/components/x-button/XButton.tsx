import { css } from '@emotion/react'
import styled from '@emotion/styled'

import type { XButtonProps } from '~/typing'

const XButton = styled.button<XButtonProps>`
  cursor: pointer;
  position: relative;
  text-align: center;
  text-transform: uppercase;

  ${props => css`
    ${props.theme.mixins.alignItems('center')};
    ${props.theme.mixins.displayFlex()};
    ${props.theme.mixins.justifyContent('center')};

    border-radius: ${props.theme.functions.spacing([
      props.theme.components.XButton.borderRadius,
    ])};

    font-size: ${props.theme.functions.spacing([
      props.theme.components.XButton.size[props.size || 'normal'].fontSize,
    ])};
    padding: ${props.theme.functions.spacing(
      props.theme.components.XButton.size[props.size || 'normal'].padding,
    )};

    height: ${props.theme.functions.spacing([
      props.theme.components.XButton.size[props.size || 'normal'].height,
    ])};
  `}

  &:before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    ${props => css`
      ${props.theme.mixins.transition(
        props.theme.functions.transition(['background-color'], 'fast'),
      )};

      border-radius: inherit;
    `}
  }

  &:hover:before {
    background-color: rgba(0, 0, 0, 0.15);
  }

  ${props => {
    switch (props.variant) {
      case 'text':
        return css`
          color: ${props.theme.components.XButton.color[props.color || 'base']};
          background-color: transparent;
        `
      case 'outline':
        return css`
          color: ${props.theme.components.XButton.color[props.color || 'base']};
          background-color: transparent;

          &:before {
            border: 1px solid
              ${props.theme.components.XButton.color[props.color || 'base']};
          }
        `
      default:
        return css`
          color: ${props.theme.components.XButton.color.contrastText};
          background-color: ${props.theme.components.XButton.color[
            props.color || 'base'
          ]};

          ${props.theme.mixins.boxShadow(props.theme.components.XButton.shadow)}

          ${((props.animation?.onHover ??
            props.theme.components?.XButton?.animation?.onHover) ||
            (props.animation?.onActive ??
              props.theme.components?.XButton?.animation?.onActive)) &&
          css`
            ${props.theme.mixins.transition(
              props.theme.functions.transition(
                props.theme.components.XButton.transition.base?.properties ||
                  [],
                props.theme.components.XButton.transition.base?.speed || 'fast',
              ),
            )}

            &:active {
              ${props.theme.mixins.transition(
                props.theme.functions.transition(
                  props.theme.components.XButton.transition.onActive
                    ?.properties || [],
                  props.theme.components.XButton.transition.onActive?.speed ||
                    'faster',
                ),
              )}
            }
          `}

          ${(props.animation?.onHover ??
            props.theme.components.XButton.animation.onHover) &&
          css`
            &:hover {
              ${props.theme.mixins.transform(
                props.theme.components.XButton.transform.onHover,
              )}
            }
          `}

          ${(props.animation?.onActive ??
            props.theme.components?.XButton?.animation?.onActive) &&
          css`
            &:active {
              ${props.theme.mixins.transform(
                props.theme.components.XButton.transform.onActive,
              )}
            }
          `}
        `
    }
  }}
`

export default XButton
