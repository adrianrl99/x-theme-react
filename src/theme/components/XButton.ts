import { baseColorPalette, baseDarkPalette } from '~/theme/base'
import { functions } from '~/theme/functions'
import { XButtonThemeProps } from '~/typing'

const XButtonThemeConfig: XButtonThemeProps = {
  animation: {
    onHover: true,
    onActive: true,
  },
  borderRadius: 2,
  color: {
    base: baseDarkPalette.white.normal,
    contrastText: baseDarkPalette.white.contrastText,
    success: baseColorPalette.success,
    info: baseColorPalette.info,
    warning: baseColorPalette.warning,
    error: baseColorPalette.error,
  },
  shadow: functions.shadow(0),
  size: {
    small: {
      fontSize: 3.5,
      padding: [1.5, 4],
      height: 7,
    },
    normal: {
      fontSize: 4,
      padding: [2, 5],
      height: 8,
    },
    large: {
      fontSize: 5,
      padding: [2.5, 6],
      height: 9,
    },
  },
  transform: {
    onHover: 'scale(1.08)',
    onActive: 'scale(0.95)',
  },
  transition: {
    base: {
      speed: 'fast',
      properties: ['transform'],
    },
    onActive: {
      speed: 'faster',
      properties: ['transform'],
    },
  },
}

export default XButtonThemeConfig
