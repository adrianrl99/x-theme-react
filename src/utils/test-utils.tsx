import '#/window.ts'

import type { RenderOptions } from '@testing-library/react'
import { render } from '@testing-library/react'
import type { PropsWithChildren, ReactElement, ReactNode } from 'react'

import { XThemeProvider } from '~/theme'

const AllTheProviders = ({ children }: PropsWithChildren<ReactNode>) => (
  <XThemeProvider>{children}</XThemeProvider>
)

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
