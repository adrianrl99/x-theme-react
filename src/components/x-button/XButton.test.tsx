import '@testing-library/jest-dom'

import { render } from 'test-utils'

import XButton from './XButton'

test('render x-button', function t() {
  const { container, getByText } = render(<XButton>button</XButton>)
  expect(getByText('button')).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})
