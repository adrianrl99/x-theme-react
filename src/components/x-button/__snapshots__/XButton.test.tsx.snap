// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render x-button 1`] = `
<div>
  <button
    class="css-1yjo7i8"
  >
    button
  </button>
</div>
`;
